File: Arrays & Hashing/0525. Contiguous Array/0525.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        prefix_sum = {0:-1}
        prefix = 0
        res = 0

        for i,v in enumerate(nums):
            prefix += 1 if v == 1 else -1
            if prefix not in prefix_sum:
                prefix_sum[prefix] = i
            else:
                res = max(res, i - prefix_sum[prefix])

        return res
            

File: Arrays & Hashing/0525. Contiguous Array/0525.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefixSum((n*2)+1,-2);
        prefixSum[n] = -1;
        int prefix = 0;
        int res = 0;

        for (int i=0; i<nums.size(); ++i) {
            prefix += nums[i] > 0 ? 1:-1;
            if (prefixSum[prefix+n]==-2) {
                prefixSum[prefix+n] = i;
            } else {
                res = max(res, i - prefixSum[prefix+n]);
            }
        } return res;
    }
};